File: ios/Classes/Scanner Core/Scan Filters/ExpiryDateFilter.swift
//
//  ExpiryDateFilter.swift
//  card_scanner
//
//  Created by Abhinav Kumar on 12/02/21.
//

import Foundation
import MLKitTextRecognition

class ExpiryDateFilter : ScanFilter {
    var visionText: Text
    var scannerOptions: CardScannerOptions
    private var cardNumberScanResult: CardNumberScanResult
    
    private var expiryDateRegex: NSRegularExpression = try! NSRegularExpression(pattern: CardScannerRegexps.expiryDateRegex, options: .anchorsMatchLines)
    
    var maxBlocksBelowCardNumberToSearchForExpiryDate = 4
    var expiryDateFormat = "MM/yy"
    
    init(visionText: Text, scannerOptions: CardScannerOptions, cardNumberScanResult: CardNumberScanResult) {
        self.visionText = visionText
        self.scannerOptions = scannerOptions
        self.cardNumberScanResult = cardNumberScanResult
    }
    
    func filter() -> ScanFilterResult? {
        guard cardNumberScanResult.cardNumber.isNotEmpty else { return nil }
        guard scannerOptions.scanExpiryDate else { return nil }
        
        var scanResults: [ExpiryDateScanResult] = []
        
        let maxTextBlockIndexToSearchExpiryDate = min(
            cardNumberScanResult.textBlockIndex + maxBlocksBelowCardNumberToSearchForExpiryDate,
            visionText.blocks.count - 1
        )
        
        for index in cardNumberScanResult.textBlockIndex...maxTextBlockIndexToSearchExpiryDate {
            let block = visionText.blocks[index]
            for line in block.lines {
                let potentialDate = line.text
                let nsString = potentialDate as NSString
                if (!(expiryDateRegex.numberOfMatches(in: line.text, range: NSRange(location: 0, length: line.text.count)) > 0)) {
                    continue
                }
                
                for match in expiryDateRegex.matches(in: potentialDate, range: NSRange(location: 0, length: potentialDate.count)) {
                    
                    let expiryDate = nsString.substring(with: match.range).trimmingCharacters(in: .whitespacesAndNewlines)
                    if (isValidExpiryDate(expiryDate: expiryDate)) {
                        scanResults.append(
                            ExpiryDateScanResult(
                                visionText: visionText,
                                textBlockIndex: index,
                                textBlock: block,
                                expiryDate: expiryDate
                            )
                        )
                    }
                }
                
                if (scanResults.count > 2) { break }
            }
        }
        
        if (scanResults.isEmpty) { return nil }
        return chooseMostRecentDate(expiryDateResults: scanResults)
    }
    
    func chooseMostRecentDate(expiryDateResults: [ExpiryDateScanResult]) -> ExpiryDateScanResult {
        if (expiryDateResults.count == 1) { return expiryDateResults[0] }
        var mostRecentDateResult = expiryDateResults[0]
        for (_, expiryDateResult) in expiryDateResults[1..<expiryDateResults.count].enumerated(){
            let currentMostRecent = parseExpiryDate(expiryDate: mostRecentDateResult.expiryDate)
            let newDate = parseExpiryDate(expiryDate: expiryDateResult.expiryDate)
            if (newDate > currentMostRecent) {
                mostRecentDateResult = expiryDateResult
            }
        }
        return mostRecentDateResult
    }
    
    func isValidExpiryDate(expiryDate: String) -> Bool{
        let dateFormatter = DateFormatter()
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        dateFormatter.dateFormat = expiryDateFormat
        
        guard let expiryDateTime = dateFormatter.date(from: expiryDate) else { return false }
        let currentDateTime = Date()
        if (scannerOptions.considerPastDatesInExpiryDateScan) {
            return true
        } else {
            return expiryDateTime > currentDateTime
        }
    }
    
    func parseExpiryDate(expiryDate: String) -> Date {
        let dateFormatter = DateFormatter()
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        dateFormatter.dateFormat = expiryDateFormat
        let date = dateFormatter.date(from: expiryDate)!
        return date
    }
}

File: android/src/main/kotlin/com/basys/card_scanner/scanner_core/scan_filters/ExpiryDateFilter.kt
package com.basys.card_scanner.scanner_core.scan_filters

import com.google.mlkit.vision.text.Text
import com.basys.card_scanner.scanner_core.constants.CardScannerRegexps
import com.basys.card_scanner.scanner_core.models.*
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.min

class ExpiryDateFilter(visionText: Text, scannerOptions: CardScannerOptions, private val cardNumberScanResult: CardNumberScanResult) : ScanFilter(visionText, scannerOptions) {
  private val expiryDateRegex: Regex = Regex(CardScannerRegexps.expiryDateRegex, RegexOption.MULTILINE)
  val _maxBlocksBelowCardNumberToSearchForExpiryDate = 4;
  val _expiryDateFormat = "MM/yy";

  override fun filter(): ExpiryDateScanResult? {
    if (cardNumberScanResult.cardNumber.isEmpty()) return null;
    if (!scannerOptions.scanExpiryDate) return null;

    val scanResults: MutableList<ExpiryDateScanResult> = mutableListOf()
    val maxTextBlockIndexToSearchExpiryDate = min(
            cardNumberScanResult.textBlockIndex + _maxBlocksBelowCardNumberToSearchForExpiryDate, visionText.textBlocks.size - 1);

    for (index in cardNumberScanResult.textBlockIndex..maxTextBlockIndexToSearchExpiryDate) {
      val block = visionText.textBlocks[index]
      if (!expiryDateRegex.containsMatchIn(block.text)) continue;
      for (match in expiryDateRegex.findAll(block.text)) {
        val expiryDate = match.groupValues[0].trim().replace(Regex("\\s+"), "")
        if (isValidExpiryDate(expiryDate)) {
          scanResults.add(ExpiryDateScanResult(
                  textBlockIndex = index, textBlock = block, expiryDate = expiryDate, visionText = visionText));
        }
      }
      if (scanResults.size > 2) break;
    }

    if (scanResults.isEmpty()) return null;
    return chooseMostRecentDate(scanResults);
  }

  private fun chooseMostRecentDate(expiryDateResults: List<ExpiryDateScanResult>): ExpiryDateScanResult {
    if (expiryDateResults.size == 1) return expiryDateResults[0];

    var mostRecentDateResult = expiryDateResults[0]
    for (expiryDateResult in expiryDateResults.subList(1, expiryDateResults.size)) {
      val currentMostRecent = parseExpiryDate(mostRecentDateResult.expiryDate);
      val newDate = parseExpiryDate(expiryDateResult.expiryDate);
      if (newDate != null) {
        if (newDate.after(currentMostRecent)) {
          mostRecentDateResult = expiryDateResult;
        }
      }
    }
    return mostRecentDateResult;
  }

  private fun isValidExpiryDate(expiryDate: String): Boolean {
    val expiryDateTime = SimpleDateFormat(_expiryDateFormat, Locale.US).parse(expiryDate)
    val currentDateTime = SimpleDateFormat(_expiryDateFormat, Locale.US).parse(
            SimpleDateFormat(_expiryDateFormat, Locale.US).format(Date()))
    return if (scannerOptions.considerPastDatesInExpiryDateScan) {
      true;
    } else {
      expiryDateTime?.after(currentDateTime) ?: false
    }
  }

  private fun parseExpiryDate(expiryDate: String): Date? {
    return SimpleDateFormat(_expiryDateFormat, Locale.US).parse(expiryDate)
  }
}
